# Segmenting and Clustering Toronto Neighbourhoods

## Obtaining the data... ##

First, import the libraries we will require in this notebook.
We use:
* <b>urllib.request</b> to open the Wikipedia URL provided (https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M)
* <b>BeautifulSoup</b> to parse our HTML file and scrape the table
* <b>pandas</b> to work with the data we have in the Wikipedia table more easily, in a dataframe

In [1]:
import urllib.request
from bs4 import BeautifulSoup
import pandas as pd

So now we can use urlopen() from the urllib.request library to open the URL provided for the Canadian Postal Codes Wikipedia page...

In [2]:
url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
page = urllib.request.urlopen(url)

... and we used BeautifulSoup to parse the HTML file into our notebook. \
<b>Note:</b> soup.prettify() formats the parse tree so that it is easier to read and work with

In [3]:
soup = BeautifulSoup(page, "lxml")
print(soup.prettify())

<!DOCTYPE html>
<html class="client-nojs" dir="ltr" lang="en">
 <head>
  <meta charset="utf-8"/>
  <title>
   List of postal codes of Canada: M - Wikipedia
  </title>
  <script>
   document.documentElement.className="client-js";RLCONF={"wgBreakFrames":!1,"wgSeparatorTransformTable":["",""],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","January","February","March","April","May","June","July","August","September","October","November","December"],"wgRequestId":"628d4537-c294-4879-8495-90635801e84b","wgCSPNonce":!1,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":!1,"wgNamespaceNumber":0,"wgPageName":"List_of_postal_codes_of_Canada:_M","wgTitle":"List of postal codes of Canada: M","wgCurRevisionId":969048979,"wgRevisionId":969048979,"wgArticleId":539066,"wgIsArticle":!0,"wgIsRedirect":!1,"wgAction":"view","wgUserName":null,"wgUserGroups":["*"],"wgCategories":["Articles with short description","Communications in Ontario","Postal codes in Canada","Toron

So in the cell above we can see we have parsed the HTML for the full Wikipedia page. We don't need the whole page - we only want to work with the 'Postal Codes' table. All tables with a HTML file are coded using the 'table' tag, so let's first find <i>all</i> tables in the Wikipedia page. As you can see in the output cell below, this gives us four tables:

In [4]:
table=soup.find_all("table")
table

[<table class="wikitable sortable">
 <tbody><tr>
 <th>Postal Code
 </th>
 <th>Borough
 </th>
 <th>Neighbourhood
 </th></tr>
 <tr>
 <td>M1A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>
 <tr>
 <td>M2A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>
 <tr>
 <td>M3A
 </td>
 <td>North York
 </td>
 <td>Parkwoods
 </td></tr>
 <tr>
 <td>M4A
 </td>
 <td>North York
 </td>
 <td>Victoria Village
 </td></tr>
 <tr>
 <td>M5A
 </td>
 <td>Downtown Toronto
 </td>
 <td>Regent Park, Harbourfront
 </td></tr>
 <tr>
 <td>M6A
 </td>
 <td>North York
 </td>
 <td>Lawrence Manor, Lawrence Heights
 </td></tr>
 <tr>
 <td>M7A
 </td>
 <td>Downtown Toronto
 </td>
 <td>Queen's Park, Ontario Provincial Government
 </td></tr>
 <tr>
 <td>M8A
 </td>
 <td>Not assigned
 </td>
 <td>Not assigned
 </td></tr>
 <tr>
 <td>M9A
 </td>
 <td>Etobicoke
 </td>
 <td>Islington Avenue, Humber Valley Village
 </td></tr>
 <tr>
 <td>M1B
 </td>
 <td>Scarborough
 </td>
 <td>Malvern, Rouge
 </td></tr>
 <tr>
 <td>M2B

The table we want to work with is the only table with class 'wikitable sortable', so let's use this distinguishing feature to save the Toronto Postal Codes table as the 'right_table'...

In [5]:
right_table=soup.find('table', class_='wikitable sortable')
right_table

<table class="wikitable sortable">
<tbody><tr>
<th>Postal Code
</th>
<th>Borough
</th>
<th>Neighbourhood
</th></tr>
<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A
</td>
<td>North York
</td>
<td>Parkwoods
</td></tr>
<tr>
<td>M4A
</td>
<td>North York
</td>
<td>Victoria Village
</td></tr>
<tr>
<td>M5A
</td>
<td>Downtown Toronto
</td>
<td>Regent Park, Harbourfront
</td></tr>
<tr>
<td>M6A
</td>
<td>North York
</td>
<td>Lawrence Manor, Lawrence Heights
</td></tr>
<tr>
<td>M7A
</td>
<td>Downtown Toronto
</td>
<td>Queen's Park, Ontario Provincial Government
</td></tr>
<tr>
<td>M8A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9A
</td>
<td>Etobicoke
</td>
<td>Islington Avenue, Humber Valley Village
</td></tr>
<tr>
<td>M1B
</td>
<td>Scarborough
</td>
<td>Malvern, Rouge
</td></tr>
<tr>
<td>M2B
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3B
</td>
<td

Perfect! Now we have to start looping through the rows to get the data for every postal code in the table. This is fairly straightforward as each postal code has it’s own defined row in the table.


We set up three empty lists (A, B and C) to store our data in, one for each column in the table. We use the Beautiful Soup ‘find_all’ function again and set it to look for the string ‘tr’, which notates the start of a new row. We will then set up a FOR loop for each row within that array and set Python to loop through the rows, one by one.

Within the loop we are going to use find_all again to search each row for td tags. We will add all of these to a variable called ‘cells’ and then check to make sure that there are 3 items in our ‘cells’ array (i.e. one for each column). If there are then we use the find(text=True)) option to extract the content string from within each <td> element in that row and add them to the A-C lists we created at the start of this step...

In [6]:
A=[]
B=[]
C=[]

for row in right_table.findAll('tr'):
    cells=row.findAll('td')
    if len(cells)==3:
        A.append(cells[0].find(text=True))
        B.append(cells[1].find(text=True))
        C.append(cells[2].find(text=True))

Now we use our lists to create a dataframe, assigning each of the lists A-C into a column with the name of our source table columns i.e Postal Code, Borough, Neighbourhood...

In [7]:
df=pd.DataFrame(A,columns=['Postal Code'])
df['Borough']=B
df['Neighbourhood']=C
df.head()

Postal Code            Borough               Neighbourhood
0        M1A
      Not assigned
               Not assigned

1        M2A
      Not assigned
               Not assigned

2        M3A
        North York
                  Parkwoods

3        M4A
        North York
           Victoria Village

4        M5A
  Downtown Toronto
  Regent Park, Harbourfront

Now we must ignore all postal codes without an assigned borough. To do this, first we should ensure the data type of values in the 'Borough' column is 'string'...

In [8]:
df['Borough']=df['Borough'].astype('str')
df.head()

Postal Code             Borough               Neighbourhood
0        M1A
      Not assigned\n               Not assigned

1        M2A
      Not assigned\n               Not assigned

2        M3A
        North York\n                  Parkwoods

3        M4A
        North York\n           Victoria Village

4        M5A
  Downtown Toronto\n  Regent Park, Harbourfront

There must have been a new line after each value in the table, notated by '\n'. We can clean our table by removing every '\n' from the Neighbourhood column...

In [9]:
df = df.replace('\n','', regex=True)
df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

That's better! Now we can drop all postal codes which have 'Not assigned' as their borough...

In [10]:
df.drop(df[df.Borough=='Not assigned'].index, axis=0, inplace=True)
df.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

If more than one neighborhood exists in one postal code area, we want to combine all rows for that postal code into one row, with the neighborhoods comma separated. We can do this using the code below...

In [11]:
df.groupby('Postal Code').agg({'Borough' : 'first', 'Neighbourhood' : ','.join})

Borough  \
Postal Code                     
M1B               Scarborough   
M1C               Scarborough   
M1E               Scarborough   
M1G               Scarborough   
M1H               Scarborough   
M1J               Scarborough   
M1K               Scarborough   
M1L               Scarborough   
M1M               Scarborough   
M1N               Scarborough   
M1P               Scarborough   
M1R               Scarborough   
M1S               Scarborough   
M1T               Scarborough   
M1V               Scarborough   
M1W               Scarborough   
M1X               Scarborough   
M2H                North York   
M2J                North York   
M2K                North York   
M2L                North York   
M2M                North York   
M2N                North York   
M2P                North York   
M2R                North York   
M3A                North York   
M3B                North York   
M3C                North York   
M3H                North York   
M3J                North York   
...                       ...   
M6C                      York   
M6E                      York   
M6G          Downtown Toronto   
M6H              West Toronto   
M6J              West Toronto   
M6K              West Toronto   
M6L                North York   
M6M                      York   
M6N                      York   
M6P              West Toronto   
M6R              West Toronto   
M6S              West Toronto   
M7A          Downtown Toronto   
M7R               Mississauga   
M7Y              East Toronto   
M8V                 Etobicoke   
M8W                 Etobicoke   
M8X                 Etobicoke   
M8Y                 Etobicoke   
M8Z                 Etobicoke   
M9A                 Etobicoke   
M9B                 Etobicoke   
M9C                 Etobicoke   
M9L                North York   
M9M                North York   
M9N                      York   
M9P                 Etobicoke   
M9R                 Etobicoke   
M9V                 Etobicoke   
M9W                 Etobicoke   

                                                 Neighbourhood  
Postal Code                                                     
M1B                                             Malvern, Rouge  
M1C                     Rouge Hill, Port Union, Highland Creek  
M1E                          Guildwood, Morningside, West Hill  
M1G                                                     Woburn  
M1H                                                  Cedarbrae  
M1J                                        Scarborough Village  
M1K                Kennedy Park, Ionview, East Birchmount Park  
M1L                            Golden Mile, Clairlea, Oakridge  
M1M            Cliffside, Cliffcrest, Scarborough Village West  
M1N                                Birch Cliff, Cliffside West  
M1P          Dorset Park, Wexford Heights, Scarborough Town...  
M1R                                          Wexford, Maryvale  
M1S                                                  Agincourt  
M1T                    Clarks Corners, Tam O'Shanter, Sullivan  
M1V          Milliken, Agincourt North, Steeles East, L'Amo...  
M1W                              Steeles West, L'Amoreaux West  
M1X                                                Upper Rouge  
M2H                                          Hillcrest Village  
M2J                               Fairview, Henry Farm, Oriole  
M2K                                            Bayview Village  
M2L                                   York Mills, Silver Hills  
M2M                                    Willowdale, Newtonbrook  
M2N                                Willowdale, Willowdale East  
M2P                                            York Mills West  
M2R                                Willowdale, Willowdale West  
M3A                                                  Parkwoods  
M3B                                                  Don Mills  
M3C                                                  Don 

We have dropped all postal codes without a borough, but let's check to see whether any neighbourhoods are left unassigned. First we should clean the 'Neighbourhood' column by converting values to type 'string' and removing all '\n's, just as we did with the Borough column...

In [12]:
df['Neighbourhood']=df['Neighbourhood'].astype('str')
df = df.replace('\n','', regex=True)
df.head()

Postal Code           Borough                                Neighbourhood
2         M3A        North York                                    Parkwoods
3         M4A        North York                             Victoria Village
4         M5A  Downtown Toronto                    Regent Park, Harbourfront
5         M6A        North York             Lawrence Manor, Lawrence Heights
6         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

Now we can see if any of the remaining postal codes in our table have 'Not assigned' as the value in the 'Neighbourhood' column...

In [13]:
df[df.Neighbourhood == 'Not assigned'].count()

Postal Code      0
Borough          0
Neighbourhood    0
dtype: int64

Perfect! Now that we have cleaned our dataframe, let's see how many rows we are left with...

In [14]:
df.shape

(103, 3)

# 103 rows and 3 columns!

We will read the following csv file, which contains a list of postal codes and the corresponsing co-ordinates, into a pandas dataframe. Our new dataframe will be called 'Coords'...

In [15]:
Coords=pd.read_csv('http://cocl.us/Geospatial_data')

In [16]:
Coords.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

Now we can use the 'set_index' function to set the 'Postal Code' column in both dataframes ('Coords' and 'df') as the index, then join the two dataframes into one.
Our new dataframe will be called 'df_Toronto' and will contain a list of all Toronto postal codes along with their corresponding borough, neighbourhoods, latitude and logitude...

In [17]:
df_Toronto=df.set_index('Postal Code').join(Coords.set_index('Postal Code'))
df_Toronto.head()

Borough                                Neighbourhood  \
Postal Code                                                                  
M3A                North York                                    Parkwoods   
M4A                North York                             Victoria Village   
M5A          Downtown Toronto                    Regent Park, Harbourfront   
M6A                North York             Lawrence Manor, Lawrence Heights   
M7A          Downtown Toronto  Queen's Park, Ontario Provincial Government   

              Latitude  Longitude  
Postal Code                        
M3A          43.753259 -79.329656  
M4A          43.725882 -79.315572  
M5A          43.654260 -79.360636  
M6A          43.718518 -79.464763  
M7A          43.662301 -79.389494

# Success!

## Now we want to explore and cluster our Toronto neighbourhoods. 
So that we can visualize our neighbourhoods on a map, we need to import some more libraries...
* We install <b>Geopy</b> and import <b>Nominatim</b> so that we can convert addresses into latitude and longitude
* We install and import <b>Folium</b> so that we can render maps
* We import <b>malplotlib</b> so that we can visualize our data
* We import <b>Kmeans</b> from the <b>sklearn</b> library so that we can cluster our neighbourhoods
* We import <b>requests</b> so that we can handle requests
* We import <b>json_normalize</b> so that we can transform JSON files into pandas dataframes
* We import <b>numpy</b> to handle data

In [18]:
!pip install geopy
from geopy.geocoders import Nominatim

In [19]:
!pip install folium
import folium

     |████████████████████████████████| 102kB 6.5MB/s ta 0:00:011


In [20]:
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
import requests
from pandas.io.json import json_normalize
import numpy as np

Now we use the geopy library to get the latitude and longitude values of Toronto, first defining our user_agent as 'toronto_explorer'...

In [21]:
address = 'Toronto, Canada'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical co-ordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical co-ordinates of Toronto are 43.6534817, -79.3839347.


## Ok, great! Now we can create a map of Toronto with the neighbourhoods superimposed on top...

In [24]:
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

for lat, lng, borough, neighbourhood in zip(df_Toronto['Latitude'], df_Toronto['Longitude'], df_Toronto['Borough'], df_Toronto['Neighbourhood']):
    label = '{}, {}'.format(neighbourhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

## Nice! Now let's explore our neighbourhoods using the Foursquare API...

In [25]:
CLIENT_ID = 'H0GR3YVC2EI2B2JKASDR0EVPDDCHOTGY3YLBQNWWL1WCKQJA' 
CLIENT_SECRET = '05MBKP205HIXRH2IUPPCNLT2XFOHFNX2VP0QLWI1JPHIVUEL'
VERSION = '20180605'

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: H0GR3YVC2EI2B2JKASDR0EVPDDCHOTGY3YLBQNWWL1WCKQJA
CLIENT_SECRET:05MBKP205HIXRH2IUPPCNLT2XFOHFNX2VP0QLWI1JPHIVUEL


Let's reset the index for our dataframe, then get the latitude and longitude of the first neighbourhood in our dataframe, which is Parkwoods...

In [26]:
df_Toronto=df_Toronto.reset_index()

In [27]:
neighbourhood_latitude = df_Toronto.loc[0, 'Latitude']
neighbourhood_longitude = df_Toronto.loc[0, 'Longitude'] 

neighbourhood_name = df_Toronto.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighbourhood_name, 
                                                               neighbourhood_latitude, 
                                                               neighbourhood_longitude))

Latitude and longitude values of Parkwoods are 43.7532586, -79.3296565.


Now let's get the top 100 venues within 800 metres of Parkwoods...

In [28]:
LIMIT=100
radius=800
url='https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(CLIENT_ID,CLIENT_SECRET,VERSION,neighbourhood_latitude,neighbourhood_longitude,radius,LIMIT)
results=requests.get(url).json()

results

{'meta': {'code': 200, 'requestId': '5f1947c2803a4b4e51034118'},
 'response': {'headerLocation': 'Parkwoods - Donalda',
  'headerFullLocation': 'Parkwoods - Donalda, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 5,
  'suggestedBounds': {'ne': {'lat': 43.76045860720001,
    'lng': -79.31970728375885},
   'sw': {'lat': 43.7460585928, 'lng': -79.33960571624115}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4e8d9dcdd5fbbbb6b3003c7b',
       'name': 'Brookbanks Park',
       'location': {'address': 'Toronto',
        'lat': 43.751976046055574,
        'lng': -79.33214044722958,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.751976046055574,
          'lng': -79.33214044722958}],
        'distance': 245,
        'cc': 'CA',

The below function extracts the category of the venue...

In [29]:
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

So that we can add a category to each of the venues in our dataframe...

In [30]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues)
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name         categories        lat        lng
0    Brookbanks Park               Park  43.751976 -79.332140
1      Variety Store  Food & Drink Shop  43.751974 -79.333114
2  DVP at York Mills               Road  43.758899 -79.334099
3    TTC Stop #09083           Bus Stop  43.759655 -79.332223
4      TTC Stop 9083           Bus Stop  43.759251 -79.334000

Now let's see how many venues are within 800m of Parkwoods...

In [92]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

5 venues were returned by Foursquare.


Let's now find the top 100 venues within 800m of all Toronto neighbourhoods...

In [32]:
def getNearbyVenues(names, latitudes, longitudes, radius=800):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighbourhood', 
                  'Neighbourhood Latitude', 
                  'Neighbourhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Then run through all neighbourhoods to create a <i>new dataframe</i> called Toronto_venues which has the co-ordinates of each venue in each Toronto neighbourhood...

In [35]:
Toronto_venues = getNearbyVenues(names=df_Toronto['Neighbourhood'],
                                   latitudes=df_Toronto['Latitude'],
                                   longitudes=df_Toronto['Longitude']
                                  )

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Queen's Park, Ontario Provincial Government
Islington Avenue, Humber Valley Village
Malvern, Rouge
Don Mills
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
East Toronto, Broadview North (Old East York)
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmo

In [36]:
print(Toronto_venues.shape)
Toronto_venues.head()

(3952, 7)


Neighbourhood  Neighbourhood Latitude  Neighbourhood Longitude  \
0     Parkwoods               43.753259               -79.329656   
1     Parkwoods               43.753259               -79.329656   
2     Parkwoods               43.753259               -79.329656   
3     Parkwoods               43.753259               -79.329656   
4     Parkwoods               43.753259               -79.329656   

               Venue  Venue Latitude  Venue Longitude     Venue Category  
0    Brookbanks Park       43.751976       -79.332140               Park  
1      Variety Store       43.751974       -79.333114  Food & Drink Shop  
2  DVP at York Mills       43.758899       -79.334099               Road  
3    TTC Stop #09083       43.759655       -79.332223           Bus Stop  
4      TTC Stop 9083       43.759251       -79.334000           Bus Stop

## So we have 3952 venues in total. Now let's see how many venues there are in each of our neighbourhoods...

In [37]:
Toronto_venues.groupby('Neighbourhood').count()

Neighbourhood Latitude  \
Neighbourhood                                                                
Agincourt                                                               19   
Alderwood, Long Branch                                                  14   
Bathurst Manor, Wilson Heights, Downsview North                         24   
Bayview Village                                                         10   
Bedford Park, Lawrence Manor East                                       40   
Berczy Park                                                            100   
Birch Cliff, Cliffside West                                              6   
Brockton, Parkdale Village, Exhibition Place                            96   
Business reply mail Processing Centre, South Ce...                      58   
CN Tower, King and Spadina, Railway Lands, Harb...                      32   
Caledonia-Fairbanks                                                     14   
Canada Post Gateway Processing Centre                                   55   
Cedarbrae                                                               23   
Central Bay Street                                                     100   
Christie                                                                62   
Church and Wellesley                                                   100   
Clarks Corners, Tam O'Shanter, Sullivan                                 22   
Cliffside, Cliffcrest, Scarborough Village West                          8   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              84   
Davisville North                                                        58   
Del Ray, Mount Dennis, Keelsdale and Silverthorn                        11   
Don Mills                                                               32   
Dorset Park, Wexford Heights, Scarborough Town ...                      19   
Downsview                                                               49   
Dufferin, Dovercourt Village                                            41   
East Toronto, Broadview North (Old East York)                           51   
Eringate, Bloordale Gardens, Old Burnhamthorpe,...                      12   
Fairview, Henry Farm, Oriole                                            38   
First Canadian Place, Underground city                                 100   
...                                                                    ...   
Rouge Hill, Port Union, Highland Creek                                   5   
Runnymede, Swansea                                                      64   
Runnymede, The Junction North                                           31   
Scarborough Village                                                      8   
South Steeles, Silverstone, Humbergate, Jamesto...                      14   
St. James Town                                                         100   
St. James Town, Cabbagetown                                             66   
Steeles West, L'Amoreaux West                                           19   
Stn A PO Boxes                                                         100   
Studio District                                                         98   
Summerhill West, Rathnelly, South Hill, Forest ...                      64   
The Annex, North Midtown, Yorkville                                     79   
The Beaches                                                             51   
The Danforth West, Riverdale                                           100   
The Kingsway, Montgomery Road, Old Mill North                           23   
Thorncliffe Park                                                        29   
Toronto Dominion Centre, Design Exchange                               100   
University of Toronto, Harbord                                         100   
Victoria Village                                                         8

Now to analyze each neighbourhood...

In [40]:
Toronto_onehot = pd.get_dummies(Toronto_venues[['Venue Category']], prefix="", prefix_sep="")

Toronto_onehot['Neighbourhood'] = Toronto_venues['Neighbourhood'] 

fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Neighbourhood  ATM  Accessories Store  Afghan Restaurant  \
0     Parkwoods    0                  0                  0   
1     Parkwoods    0                  0                  0   
2     Parkwoods    0                  0                  0   
3     Parkwoods    0                  0                  0   
4     Parkwoods    0                  0                  0   

   African Restaurant  Airport  Airport Food Court  Airport Gate  \
0                   0        0                   0             0   
1                   0        0                   0             0   
2                   0        0                   0             0   
3                   0        0                   0             0   
4                   0        0                   0             0   

   Airport Lounge  Airport Service  ...  Video Game Store  Video Store  \
0               0                0  ...                 0            0   
1               0                0  ...                 0            0   
2               0                0  ...                 0            0   
3               0                0  ...                 0            0   
4               0                0  ...                 0            0   

   Vietnamese Restaurant  Warehouse Store  Whisky Bar  Wine Bar  Wine Shop  \
0                      0                0           0         0          0   
1                      0                0           0         0          0   
2                      0                0           0         0          0   
3                      0                0           0         0          0   
4                      0                0           0         0          0   

   Wings Joint  Women's Store  Yoga Studio  
0            0              0            0  
1            0              0            0  
2            0              0            0  
3            0              0            0  
4            0              0            0  

[5 rows x 330 columns]

In [41]:
Toronto_onehot.shape

(3952, 330)

Next, let's group rows by neighborhood and take the mean of the frequency of occurrence of each category...

In [42]:
Toronto_grouped = Toronto_onehot.groupby('Neighbourhood').mean().reset_index()
Toronto_grouped

Neighbourhood       ATM  \
0                                           Agincourt  0.000000   
1                              Alderwood, Long Branch  0.000000   
2     Bathurst Manor, Wilson Heights, Downsview North  0.000000   
3                                     Bayview Village  0.000000   
4                   Bedford Park, Lawrence Manor East  0.000000   
5                                         Berczy Park  0.000000   
6                         Birch Cliff, Cliffside West  0.000000   
7        Brockton, Parkdale Village, Exhibition Place  0.000000   
8   Business reply mail Processing Centre, South C...  0.000000   
9   CN Tower, King and Spadina, Railway Lands, Har...  0.000000   
10                                Caledonia-Fairbanks  0.000000   
11              Canada Post Gateway Processing Centre  0.000000   
12                                          Cedarbrae  0.000000   
13                                 Central Bay Street  0.000000   
14                                           Christie  0.000000   
15                               Church and Wellesley  0.000000   
16            Clarks Corners, Tam O'Shanter, Sullivan  0.045455   
17    Cliffside, Cliffcrest, Scarborough Village West  0.000000   
18                     Commerce Court, Victoria Hotel  0.000000   
19                                         Davisville  0.000000   
20                                   Davisville North  0.000000   
21   Del Ray, Mount Dennis, Keelsdale and Silverthorn  0.000000   
22                                          Don Mills  0.000000   
23  Dorset Park, Wexford Heights, Scarborough Town...  0.000000   
24                                          Downsview  0.000000   
25                       Dufferin, Dovercourt Village  0.000000   
26      East Toronto, Broadview North (Old East York)  0.000000   
27  Eringate, Bloordale Gardens, Old Burnhamthorpe...  0.000000   
28                       Fairview, Henry Farm, Oriole  0.000000   
29             First Canadian Place, Underground city  0.000000   
..                                                ...       ...   
68             Rouge Hill, Port Union, Highland Creek  0.000000   
69                                 Runnymede, Swansea  0.000000   
70                      Runnymede, The Junction North  0.000000   
71                                Scarborough Village  0.000000   
72  South Steeles, Silverstone, Humbergate, Jamest...  0.000000   
73                                     St. James Town  0.000000   
74                        St. James Town, Cabbagetown  0.000000   
75                      Steeles West, L'Amoreaux West  0.000000   
76                                     Stn A PO Boxes  0.000000   
77                                    Studio District  0.000000   
78  Summerhill West, Rathnelly, South Hill, Forest...  0.000000   
79                The Annex, North Midtown, Yorkville  0.000000   
80                                        The Beaches  0.000000   
81                       The Danforth West, Riverdale  0.000000   
82      The Kingsway, Montgomery Road, Old Mill North  0.000000   
83                                   Thorncliffe Park  0.000000   
84           Toronto Dominion Centre, Design Exchange  0.000000   
85                     University of Toronto, Harbord  0.000000   
86                                   Victoria Village  0.000000   
87  West Deane Park, Princess Gardens, Martin Grov...  0.000000   
88                                          Westmount  0.000000   
89                                             Weston  0.000000   
90                                  Wexford, Maryvale  0.000000   
91                            Willowdale, Newtonbrook  0.000000   
92                        Willowdale, Willowdale East  0.000000   
93                        Willowdale, Willowdale West  0.000000   
94                                             Woburn  0.000000   
95                                   Woodbine Heights  0.000000   
96                  

The following functions sorts the venue categories into descending order...

In [44]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

... so that we can put this information into a dataframe (top 10 venue categories in each neighbourhood)...

In [219]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighbourhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighbourhoods_venues_sorted = pd.DataFrame(columns=columns)
neighbourhoods_venues_sorted['Neighbourhood'] = Toronto_grouped['Neighbourhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighbourhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighbourhoods_venues_sorted.head()

Neighbourhood 1st Most Common Venue  \
0                                        Agincourt    Chinese Restaurant   
1                           Alderwood, Long Branch     Convenience Store   
2  Bathurst Manor, Wilson Heights, Downsview North           Pizza Place   
3                                  Bayview Village   Japanese Restaurant   
4                Bedford Park, Lawrence Manor East           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0           Fabric Shop                Lounge        Sandwich Place   
1           Pizza Place                  Park           Coffee Shop   
2           Coffee Shop                  Bank                  Park   
3                  Bank          Skating Rink               Dog Run   
4    Italian Restaurant           Pizza Place        Cosmetics Shop   

  5th Most Common Venue     6th Most Common Venue      7th Most Common Venue  \
0        Discount Store        Seafood Restaurant  Latin American Restaurant   
1                   Pub                       Gym             Sandwich Place   
2        Ice Cream Shop  Mediterranean Restaurant  Middle Eastern Restaurant   
3                  Park             Grocery Store         Chinese Restaurant   
4            Restaurant            Sandwich Place              Grocery Store   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0      Sushi Restaurant       Motorcycle Shop            Supermarket  
1        Discount Store              Pharmacy             Donut Shop  
2     Mobile Phone Shop           Bridal Shop            Supermarket  
3                  Café            Donut Shop         Discount Store  
4             Locksmith            Bagel Shop                 Bakery

## Ok, now we can use this information to cluster our neighbourhoods by venue category...

In [220]:
kclusters = 7

Toronto_grouped_clustering = Toronto_grouped.drop('Neighbourhood', 1)

kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

kmeans.labels_[0:10] 

array([1, 2, 2, 0, 1, 1, 1, 1, 1, 1], dtype=int32)

Now we can add this to our dataframe...

In [221]:
neighbourhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

neighbourhoods_venues_sorted.head()

Cluster Labels                                    Neighbourhood  \
0               1                                        Agincourt   
1               2                           Alderwood, Long Branch   
2               2  Bathurst Manor, Wilson Heights, Downsview North   
3               0                                  Bayview Village   
4               1                Bedford Park, Lawrence Manor East   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0    Chinese Restaurant           Fabric Shop                Lounge   
1     Convenience Store           Pizza Place                  Park   
2           Pizza Place           Coffee Shop                  Bank   
3   Japanese Restaurant                  Bank          Skating Rink   
4           Coffee Shop    Italian Restaurant           Pizza Place   

  4th Most Common Venue 5th Most Common Venue     6th Most Common Venue  \
0        Sandwich Place        Discount Store        Seafood Restaurant   
1           Coffee Shop                   Pub                       Gym   
2                  Park        Ice Cream Shop  Mediterranean Restaurant   
3               Dog Run                  Park             Grocery Store   
4        Cosmetics Shop            Restaurant            Sandwich Place   

       7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0  Latin American Restaurant      Sushi Restaurant       Motorcycle Shop   
1             Sandwich Place        Discount Store              Pharmacy   
2  Middle Eastern Restaurant     Mobile Phone Shop           Bridal Shop   
3         Chinese Restaurant                  Café            Donut Shop   
4              Grocery Store             Locksmith            Bagel Shop   

  10th Most Common Venue  
0            Supermarket  
1             Donut Shop  
2            Supermarket  
3         Discount Store  
4                 Bakery

Now we want to merge this dataframe with the original dataframe we created containing geographical co-ordinates and borough information, called df_Toronto. In order to join the two dataframes together, both must have the 'Neighbourhood' column set as the index. 

In [222]:
Toronto_merged = df_Toronto.join(neighbourhoods_venues_sorted.set_index('Neighbourhood'), on='Neighbourhood')

In [223]:
Toronto_merged.head()

index Postal Code           Borough  \
0      0         M3A        North York   
1      1         M4A        North York   
2      2         M5A  Downtown Toronto   
3      3         M6A        North York   
4      4         M7A  Downtown Toronto   

                                 Neighbourhood   Latitude  Longitude  \
0                                    Parkwoods  43.753259 -79.329656   
1                             Victoria Village  43.725882 -79.315572   
2                    Regent Park, Harbourfront  43.654260 -79.360636   
3             Lawrence Manor, Lawrence Heights  43.718518 -79.464763   
4  Queen's Park, Ontario Provincial Government  43.662301 -79.389494   

   Cluster Labels   1st Most Common Venue 2nd Most Common Venue  \
0             0.0                Bus Stop                  Road   
1             1.0   Portuguese Restaurant           Coffee Shop   
2             1.0             Coffee Shop                  Park   
3             1.0  Furniture / Home Store  Fast Food Restaurant   
4             1.0             Coffee Shop    Italian Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Park     Food & Drink Shop   Dumpling Restaurant   
1   Sporting Goods Shop     French Restaurant                  Park   
2                Bakery                  Café                   Pub   
3   Fried Chicken Joint          Dessert Shop        Clothing Store   
4       Bubble Tea Shop             Bookstore             Gastropub   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Discount Store   Distribution Center              Dive Bar   
1            Pizza Place                  Café          Hockey Arena   
2                Theater            Restaurant       Thai Restaurant   
3  Vietnamese Restaurant           Coffee Shop    Athletics & Sports   
4                   Park     French Restaurant   Japanese Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0               Dog Run       Doner Restaurant  
1           Fabric Shop                  Diner  
2        Breakfast Spot  Performing Arts Venue  
3                  Café    Arts & Crafts Store  
4        Ice Cream Shop       Sushi Restaurant

In Toronto_merged, we can see that the data type of the 'Cluster Labels' values has changed from integer to float. If we want to use the cluster labels to categorize our clusters then they should be integers, however we aren't able to convert the values in the 'Cluster Labels' column from Float to Integer as there is one NaN values ('Upper Rouge' has not been assigned a cluster as there are no venues with 800m).\
Let's drop those rows...

In [224]:
Toronto_merged.isna().sum()

index                     0
Postal Code               0
Borough                   0
Neighbourhood             0
Latitude                  0
Longitude                 0
Cluster Labels            1
1st Most Common Venue     1
2nd Most Common Venue     1
3rd Most Common Venue     1
4th Most Common Venue     1
5th Most Common Venue     1
6th Most Common Venue     1
7th Most Common Venue     1
8th Most Common Venue     1
9th Most Common Venue     1
10th Most Common Venue    1
dtype: int64

In [225]:
Toronto_merged=Toronto_merged.dropna()
Toronto_merged['Cluster Labels']=Toronto_merged['Cluster Labels'].astype(int)
Toronto_merged.head()

index Postal Code           Borough  \
0      0         M3A        North York   
1      1         M4A        North York   
2      2         M5A  Downtown Toronto   
3      3         M6A        North York   
4      4         M7A  Downtown Toronto   

                                 Neighbourhood   Latitude  Longitude  \
0                                    Parkwoods  43.753259 -79.329656   
1                             Victoria Village  43.725882 -79.315572   
2                    Regent Park, Harbourfront  43.654260 -79.360636   
3             Lawrence Manor, Lawrence Heights  43.718518 -79.464763   
4  Queen's Park, Ontario Provincial Government  43.662301 -79.389494   

   Cluster Labels   1st Most Common Venue 2nd Most Common Venue  \
0               0                Bus Stop                  Road   
1               1   Portuguese Restaurant           Coffee Shop   
2               1             Coffee Shop                  Park   
3               1  Furniture / Home Store  Fast Food Restaurant   
4               1             Coffee Shop    Italian Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                  Park     Food & Drink Shop   Dumpling Restaurant   
1   Sporting Goods Shop     French Restaurant                  Park   
2                Bakery                  Café                   Pub   
3   Fried Chicken Joint          Dessert Shop        Clothing Store   
4       Bubble Tea Shop             Bookstore             Gastropub   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0         Discount Store   Distribution Center              Dive Bar   
1            Pizza Place                  Café          Hockey Arena   
2                Theater            Restaurant       Thai Restaurant   
3  Vietnamese Restaurant           Coffee Shop    Athletics & Sports   
4                   Park     French Restaurant   Japanese Restaurant   

  9th Most Common Venue 10th Most Common Venue  
0               Dog Run       Doner Restaurant  
1           Fabric Shop                  Diner  
2        Breakfast Spot  Performing Arts Venue  
3                  Café    Arts & Crafts Store  
4        Ice Cream Shop       Sushi Restaurant

## Perfect! Now we can plot the neghbourhood clusters onto our map of Toronto...

In [226]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged['Latitude'], Toronto_merged['Longitude'], Toronto_merged['Neighbourhood'], Toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
    
map_clusters

## Finally let's review the venues we have assigned to each cluster...

<b>Cluster 0</b> is represented by blue dots. Neighbourhoods in this cluster are slightly out of town with parks, trails, leisure facilities and some grocery stores.

In [227]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 0, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Postal Code  Longitude  Cluster Labels 1st Most Common Venue  \
0           M3A -79.329656               0              Bus Stop   
21          M6E -79.453512               0              Pharmacy   
39          M2K -79.385975               0   Japanese Restaurant   
57          M9M -79.532242               0           Gas Station   
66          M2P -79.400049               0                  Park   
68          M5P -79.411307               0    Italian Restaurant   
83          M4T -79.383160               0                  Park   
91          M4W -79.377529               0                 Trail   
101         M8Y -79.498509               0         Poutine Place   

          2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                          Road                  Park     Food & Drink Shop   
21                         Café         Grocery Store                  Park   
39                         Bank          Skating Rink               Dog Run   
57            Convenience Store        Discount Store    Italian Restaurant   
66                 Tennis Court          Intersection     Convenience Store   
68                         Park      Sushi Restaurant            Bagel Shop   
83                Grocery Store           Candy Store                   Gym   
91                         Park            Playground           Candy Store   
101  Construction & Landscaping                  Park      Business Service   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0     Dumpling Restaurant        Discount Store   Distribution Center   
21    Japanese Restaurant            Beer Store        Cosmetics Shop   
39                   Park         Grocery Store    Chinese Restaurant   
57                   Park        Baseball Field          Food Service   
66             Restaurant             Pet Store           Yoga Studio   
68                 Bakery                  Bank   Japanese Restaurant   
83         Sandwich Place   Japanese Restaurant                  Café   
91          Grocery Store                  Bank           Yoga Studio   
101    Falafel Restaurant      Doner Restaurant                 Diner   

    8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Dive Bar               Dog Run       Doner Restaurant  
21     Mexican Restaurant                   Gym                   Bank  
39                   Café            Donut Shop         Discount Store  
57         Shipping Store  Ethiopian Restaurant            Event Space  
66             Donut Shop        Discount Store    Distribution Center  
68            Coffee Shop              Pharmacy              Gastropub  
83        Thai Restaurant  Gym / Fitness Center             Playground  
91             Donut Shop        Discount Store    Distribution Center  
101        Discount Store   Distribution Center         Farmers Market

<b>Cluster 1</b> is represented by purple dots on the map. Neighbourhoods in this cluster are close to the centre town with lots of coffee shops, restaurants, eateries, banks and hotels.

In [228]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 1, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Postal Code  Longitude  Cluster Labels   1st Most Common Venue  \
1           M4A -79.315572               1   Portuguese Restaurant   
2           M5A -79.360636               1             Coffee Shop   
3           M6A -79.464763               1  Furniture / Home Store   
4           M7A -79.389494               1             Coffee Shop   
6           M1B -79.194353               1    Fast Food Restaurant   
7           M3B -79.352188               1     Japanese Restaurant   
9           M5B -79.378937               1             Coffee Shop   
11          M9B -79.554724               1             Pizza Place   
12          M1C -79.160497               1          Breakfast Spot   
13          M3C -79.340923               1     Japanese Restaurant   
15          M5C -79.375418               1                    Café   
17          M9C -79.577201               1             Coffee Shop   
19          M4E -79.293031               1                     Pub   
20          M5E -79.373306               1             Coffee Shop   
23          M4G -79.363452               1             Coffee Shop   
24          M5G -79.387383               1             Coffee Shop   
25          M6G -79.422564               1           Grocery Store   
26          M1H -79.239476               1       Indian Restaurant   
30          M5H -79.384568               1             Coffee Shop   
31          M6H -79.442259               1             Coffee Shop   
33          M2J -79.346556               1             Coffee Shop   
34          M3J -79.487262               1  Furniture / Home Store   
35          M4J -79.338106               1                    Café   
36          M5J -79.381752               1             Coffee Shop   
37          M6J -79.419750               1                    Café   
38          M1K -79.262029               1             Coffee Shop   
41          M4K -79.352188               1        Greek Restaurant   
42          M5K -79.381576               1                   Hotel   
43          M6K -79.428191               1             Coffee Shop   
44          M1L -79.284577               1            Intersection   
..          ...        ...             ...                     ...   
54          M4M -79.340923               1                    Café   
55          M5M -79.419750               1             Coffee Shop   
56          M6M -79.476013               1          Discount Store   
58          M1N -79.264848               1         College Stadium   
59          M2N -79.408493               1             Pizza Place   
61          M4N -79.388790               1               Bookstore   
62          M5N -79.416936               1       Health Food Store   
65          M1P -79.273304               1  Furniture / Home Store   
67          M4P -79.390197               1             Coffee Shop   
69          M6P -79.464763               1                    Café   
71          M1R -79.295849               1             Pizza Place   
73          M4R -79.405678               1             Coffee Shop   
74          M5R -79.405678               1      Italian Restaurant   
75          M6R -79.456325               1                     Bar   
76          M7R -79.615819               1             Coffee Shop   
78          M1S -79.262029               1      Chinese Restaurant   
79          M4S -79.388790               1      Italian Restaurant   
80          M5S -79.400049               1                    Café   
81          M6S -79.484450               1             Coffee Shop   
84          M5T -79.400049               1                    Café   
86          M4V -79.400049               1             Coffee Shop   
87          M5V -79.394420               1         Harbor / Marina   
88          M8V -79.501321               1           Grocery Store   
92          M5W -79.374846               1             Coffee Shop   
96          M4X -79.367675               1             Coffee Shop   
97          M5

<b>Cluster 2</b> is represented by red dots. Neighbourhoods in this cluster are further out of town with takeaways, gas stations, and larger commercial buildings such as grocery stores and shopping malls.

In [229]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 2, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Postal Code  Longitude  Cluster Labels  1st Most Common Venue  \
5          M9A -79.532242               2               Pharmacy   
8          M4B -79.309937               2   Fast Food Restaurant   
10         M6B -79.445073               2          Grocery Store   
14         M4C -79.318389               2                   Park   
16         M6C -79.428191               2            Pizza Place   
18         M1E -79.188711               2            Pizza Place   
27         M2H -79.363452               2               Pharmacy   
28         M3H -79.442259               2            Pizza Place   
29         M4H -79.349372               2      Indian Restaurant   
32         M1J -79.239476               2         Ice Cream Shop   
40         M3K -79.464763               2  Vietnamese Restaurant   
46         M3L -79.506944               2  Vietnamese Restaurant   
51         M1M -79.239476               2            Pizza Place   
52         M2M -79.408493               2      Korean Restaurant   
53         M3M -79.495697               2  Vietnamese Restaurant   
60         M3N -79.520999               2  Vietnamese Restaurant   
63         M6N -79.487262               2                Brewery   
64         M9N -79.518188               2    Fried Chicken Joint   
70         M9P -79.532242               2            Pizza Place   
72         M2R -79.442259               2            Coffee Shop   
77         M9R -79.554724               2            Pizza Place   
82         M1T -79.304302               2            Pizza Place   
85         M1V -79.284577               2               Pharmacy   
89         M9V -79.588437               2            Pizza Place   
90         M1W -79.318389               2          Grocery Store   
93         M8W -79.543484               2      Convenience Store   

        2nd Most Common Venue 3rd Most Common Venue   4th Most Common Venue  \
5                        Park            Playground                    Café   
8                 Pizza Place               Brewery               Pet Store   
10       Fast Food Restaurant           Pizza Place             Coffee Shop   
14                   Pharmacy              Bus Stop      Athletics & Sports   
16              Grocery Store                 Field                    Park   
18                 Restaurant  Fast Food Restaurant             Coffee Shop   
27                       Park           Coffee Shop          Ice Cream Shop   
28                Coffee Shop                  Bank                    Park   
29                Coffee Shop        Sandwich Place      Turkish Restaurant   
32                Coffee Shop  Fast Food Restaurant          Sandwich Place   
40                       Park           Gas Station           Grocery Store   
46                       Park           Gas Station           Grocery Store   
51             Ice Cream Shop              Pharmacy          Hardware Store   
52                       Park           Coffee Shop          Sandwich Place   
53                       Park           Gas Station           Grocery Store   
60                       Park           Gas Station           Grocery Store   
63                Coffee Shop                  Park  Furniture / Home Store   
64                Pizza Place                  Bank       Convenience Store   
70  Middle Eastern Restaurant           Golf Course          Sandwich Place   
72                Pizza Place                  Bank           Grocery Store   
77        American Restaurant                  Bank             Gas Station   
82                   Pharmacy     Convenience Store            Intersection   
85                Pizza Place                  Park      Chinese Restaurant   
89              Grocery Store   Fried Chicken Joint          Hardware Store   
90         Chinese Restaurant           Pizza Place    Fast Food Restaurant   
93                Pizza Place                  Park             Coffee Shop   

   5th Most Common Venue        

The remaining clusters are further out of town and are not as vastly populated...

In [230]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 3, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Postal Code  Longitude  Cluster Labels 1st Most Common Venue  \
50         M9L -79.565963               3                Bakery   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
50   Arts & Crafts Store           Pizza Place           Yoga Studio   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
50  Eastern European Restaurant   Distribution Center              Dive Bar   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
50               Dog Run      Doner Restaurant             Donut Shop

In [231]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 4, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Postal Code  Longitude  Cluster Labels 1st Most Common Venue  \
45         M2L -79.374714               4                  Pool   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
45           Yoga Studio     Electronics Store        Discount Store   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
45   Distribution Center              Dive Bar               Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
45      Doner Restaurant            Donut Shop    Dumpling Restaurant

In [232]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 5, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Postal Code  Longitude  Cluster Labels 1st Most Common Venue  \
94         M9W -79.594054               5   Rental Car Location   

   2nd Most Common Venue        3rd Most Common Venue 4th Most Common Venue  \
94           Yoga Studio  Eastern European Restaurant                 Diner   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
94        Discount Store   Distribution Center              Dive Bar   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
94               Dog Run      Doner Restaurant             Donut Shop

In [233]:
Toronto_merged.loc[Toronto_merged['Cluster Labels'] == 6, Toronto_merged.columns[[1] + list(range(5, Toronto_merged.shape[1]))]]

Postal Code  Longitude  Cluster Labels 1st Most Common Venue  \
22         M1G -79.216917               6           Coffee Shop   
49         M6L -79.490074               6                  Park   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
22                  Park      Business Service            Convenience Store   
49                Bakery           Yoga Studio  Eastern European Restaurant   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
22  Eastern European Restaurant        Discount Store   Distribution Center   
49          Distribution Center              Dive Bar               Dog Run   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22              Dive Bar               Dog Run       Doner Restaurant  
49      Doner Restaurant            Donut Shop    Dumpling Restaurant

# Thank you for reviewing this notebook!